cdap-data-fabric/src/test/java/co/cask/cdap/data2/dataset2/lib/partitioned/PartitionedFileSetTest.java

```diff
@@ -547,14 +554,131 @@ public void apply() throws Exception {
+    PartitionedFileSet pfs = dsFrameworkUtil.getInstance(pfsInstance, args);
+    TransactionContext txContext = new TransactionContext(txClient, (TransactionAware) pfs);
+
+    txContext.start();
+
+    Location outputLocation = pfs.getEmbeddedFileSet().getOutputLocation();
+    Assert.assertFalse(outputLocation.exists());
+
+    // this will create the file
+    outputLocation.getOutputStream().close();
+    Assert.assertTrue(outputLocation.exists());
+
+    ((PartitionedFileSetDataset) pfs).onFailure();
+    txContext.abort();
+
+    // because the previous transaction aborted, the partition as well as the file will not exist
+    txContext.start();
+    Assert.assertNull(pfs.getPartition(PARTITION_KEY));
+    Assert.assertFalse(outputLocation.exists());
+    txContext.finish();
+  }
+
+  @Test
+  public void testRollbackOfPartitionDelete() throws Exception {
+    PartitionedFileSet pfs = dsFrameworkUtil.getInstance(pfsInstance);
+    TransactionContext txContext = new TransactionContext(txClient, (TransactionAware) pfs);
+
+    txContext.start();
+
+    PartitionOutput output = pfs.getPartitionOutput(PARTITION_KEY);
+    Location outputLocation = output.getLocation().append("file");
+    Assert.assertFalse(outputLocation.exists());
+
+    try (OutputStream outputStream = outputLocation.getOutputStream()) {
+      // write 1 to the first file
+      outputStream.write(1);
+    }
+    Assert.assertTrue(outputLocation.exists());
+
+    output.addPartition();
+    Assert.assertNotNull(pfs.getPartition(PARTITION_KEY));
+    Assert.assertTrue(pfs.getPartition(PARTITION_KEY).getLocation().exists());
+
+    txContext.finish();
+
+    // because the previous transaction aborted, the partition as well as the file will not exist
+    txContext.start();
+    pfs.dropPartition(PARTITION_KEY);
+
+    Assert.assertNull(pfs.getPartition(PARTITION_KEY));
+    Assert.assertFalse(outputLocation.exists());
+
+    // create a new partition with the same partition key (same relative path for the partition
+    PartitionOutput outputPartition2 = pfs.getPartitionOutput(PARTITION_KEY);
+    Location outputLocation2 = outputPartition2.getLocation().append("file");
+    Assert.assertFalse(outputLocation2.exists());
+    // create the file
+    try (OutputStream outputStream = outputLocation2.getOutputStream()) {
+      // write 2 to the second file
+      outputStream.write(2);
+    }
+    Assert.assertTrue(outputLocation2.exists());
+
+    outputPartition2.addPartition();
+
+    txContext.abort();
+
+    // since the previous transaction aborted, the partition and its files should still exist
+    txContext.start();
+    Assert.assertNotNull(pfs.getPartition(PARTITION_KEY));
+    Assert.assertTrue(outputLocation.exists());
+    try (InputStream inputStream = outputLocation.getInputStream()) {
+      // should be 1, written by the first partition, not 2 (which was written by the second partition)
```

Comment:
Well, I meant the first partition that was created, but if `first tx` is clearer, I can change it to that.